### PR TITLE
FEATURE: Allow NodeTypes to be defined in hierarchical folders

### PR DESCRIPTION
**Pull request by bwaidelich**
This feature allows `NodeType.*.yaml` files to be put into nested
folders underneath `<Package-Root>/NodeTypes`:

```
Packages
  Some.Package
    Configuration
      NodeTypes.yaml (still supported!)
    NodeTypes
      Foo
        Bar.yaml (new)
```

Resolves: #3000

### CODE REVIEW

**Comment by bwaidelich** [comment]
**NOTE:** This is a draft still because it relies on https://github.com/neos/flow-development-collection/pull/2449 and is missing tests still

**Comment by mficzel** [comment]
Not sure we should rush this, i would love to discuss first wether there is a way to colocate NodeTypes with the fusion integration or vice versa.

**Comment by gjwnc** [comment]
@mficzel As a note, the missing PR, @bwaidelich mentioned, allows you to put the Configuration anywhere you want. Currently, flow has the configuration locations kinda hardcoded because of the currently implemented configuration processors, tells the configuration manager where the config is located and how it should be merged.

**Comment by bwaidelich** [comment]
> we should at least think about this before.

Absolutely! I replied in the issue (https://github.com/neos/neos-development-collection/issues/3000#issuecomment-820476124)

**Comment by Sebobo** [comment]
Tested with https://github.com/neos/flow-development-collection/pull/2449 and it works perfectly! ❤️

**Review by Sebobo — CHANGES_REQUESTED** [review]
Approved but we can only merge it when https://github.com/neos/flow-development-collection/pull/2449 is merged.

**Review by jonnitto — APPROVED** [review]
Tested it, and it works ❤️

File: Neos.ContentRepository/Classes/Configuration/NodeTypesLoader.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        $this->yamlSource = $yamlSource;
+    }
+
+    /**
+     * @param PackageInterface[] $packages
+     * @param ApplicationContext $context
+     * @return array
+     * @throws ParseErrorException | ConfigurationException
+     */
+    public function __invoke(array $packages, ApplicationContext $context): array
```
Comment (bwaidelich) [inline Neos.ContentRepository/Classes/Configuration/NodeTypesLoader.php:33]:
```suggestion
    public function load(array $packages, ApplicationContext $context): array
```

**Comment by bwaidelich** [comment]
@Sebobo would you revoke your -1 now that neos/flow-development-collection#2449 is merged?

Regarding the tests: They fail because they still pull in an old version of Flow apparently (`Locking neos/flow-development-collection (7.1.x-dev 5358d96)`). @albe do you have an idea on how to solve that?

**Review by Sebobo — APPROVED** [review]

**Comment by albe** [comment]
> Regarding the tests: They fail because they still pull in an old version of Flow apparently (`Locking neos/flow-development-collection (7.1.x-dev 5358d96)`). @albe do you have an idea on how to solve that?

Very odd - the neos dev distribution is installed witth the proper dev-master and in turn requires `flow-development-collection @dev` (https://github.com/neos/neos-development-distribution/blob/master/composer.json#L20) - why composer resolves that to 7.1.x-dev instead of dev-master is not clear to me. Maybe the solution would be to also clone the flow dev-collection in the build pipeline and add a local path repository like for the neos dev-collection. That would make sure that always the matching flow branch is used in builds independent of what composer thinks it should resolve to. The biggest issue is the version mismatch before "Seven", which would require some odd logic, but maybe it's good enough to adjust the build in 7.0+ only?

See https://github.com/neos/neos-development-collection/blob/master/.github/workflows/build.yml#L115-L116

**Comment by kdambekalns** [comment]
Probably 7.1.x-dev is chosen over dev-master because it's the "better" match during dependency resolving. Maybe even our dreaded guzzlehttp/psr7 issue?

**Review by kdambekalns — APPROVED** [review]
Looks good by reading.

**Pull request merged**